### PR TITLE
FabricModule: wrap forward methods instead of monkeypatch-based redirect

### LINKED ISSUE #20710: `mark_forward_method` does not work with `ModelParallelStrategy`
### Bug description

When using the `ModelParallelStrategy`, methods annotated with `mark_forward_method` raise an exception if the function signature does not match that of the module's `forward` method. This fails specifically when the number of args/kwargs differ between the functions.

For calling `generate` here would fail in an FSDP2 setting.
```
class Model(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x, y):
        return x

    def generate(self, x, y, cfg: int = 0.5):
        z_1 = self.forward(x, y)
        z_2 = self.foward(x, torch.zeros_like(y))
        ...
```

### What version are you seeing the problem on?

v2.5

### How to reproduce the bug

```python

```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version: 2.5.0.post
#- PyTorch Version: 2.6.0+cu124
#- Python version: 3.11
#- OS: Linux
#- CUDA/cuDNN version: 12.4
#- GPU models and configuration: 8xH100
#- How you installed Lightning(`conda`, `pip`, source): pip
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by tonyf**
## What does this PR do?

When running with `torch.compile` and `ModelParallelStrategy` the monkeypatch method of redirecting forward methods through the FabricModule's custom forward method raises an exception if the marked method does not have the same signature as the original forward method.

Instead of monkeypatching forward to the method being called, this PR wraps the function to ensure the right strategy contexts are provided.

This could be cleaned up further by just wrapping the functions in `mark_forward_method` but that would require touching `_LIGHTNING_MODULE_STEP_METHODS = ("training_step", "validation_step", "test_step", "predict_step")`.

Fixes #20710

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
seems the falling test is a false alarm :)

**Comment by tonyf** [comment]
Ah! looks like I need to update the redirection tests. Will try to get around to it soon. Currently using this fix in prod successfully

**Comment by deependujha** [comment]
Hi @tonyf ,
Could you please take a look and fix failing tests so we can proceed with the merge?

**Review by Borda — COMMENTED** [review]
could you pls check the failing tests?
